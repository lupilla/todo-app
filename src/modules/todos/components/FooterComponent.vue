<template>
  <footer class="footer" v-if="getAllTodos">
    <div class="todo-count">
      <strong v-if="getActiveTodos === 1">
        {{ getActiveTodos }} item left
      </strong>
      <strong v-else> {{ getActiveTodos }} items left </strong>
    </div>
    <ul class="filters">
      <li>
        <router-link
          :data-id-test="ALL"
          to="/"
          @click="selectedEvent(ALL)"
          :class="{ selected: linkSelected === ALL }"
        >
          All
        </router-link>
        <router-link
          :data-id-test="ACTIVE"
          :to="`/${ACTIVE}`"
          @click="selectedEvent(ACTIVE)"
          :class="{ selected: linkSelected === ACTIVE }"
        >
          Active
        </router-link>
        <router-link
          :data-id-test="COMPLETED"
          :to="`/${COMPLETED}`"
          @click="selectedEvent(COMPLETED)"
          :class="{ selected: linkSelected === COMPLETED }"
        >
          Completed
        </router-link>
      </li>
    </ul>
    <button
      type="button"
      data-id-test="remove"
      class="clear-completed"
      @click="removeTodos"
      v-if="getCompletedTodos"
    >
      Clear completed
    </button>
  </footer>
</template>
<script>
import { ALL, COMPLETED, ACTIVE } from "@/modules/todos/helpers/constants.js";
import { mapGetters, mapActions } from "vuex";
export default {
  name: "FooterComponent",
  data() {
    return {
      linkSelected: ALL,
    };
  },
  props: {
    todos: {
      type: Array,
      default: () => [],
    },
  },
  computed: {
    ...mapGetters("todos", ["allTodos", "activeTodos", "completedTodos"]),
    getActiveTodos() {
      return this.activeTodos?.length;
    },
    getCompletedTodos() {
      return this.completedTodos?.length;
    },
    getAllTodos() {
      return this.allTodos?.length;
    },
  },
  methods: {
    ...mapActions("todos", ["deleteCompletedTodos"]),
    async removeTodos() {
      try {
        await this.deleteCompletedTodos();
      } catch (error) {
        console.error(error);
      }
    },
    selectedEvent(todoLink) {
      this.linkSelected = todoLink;
    },
  },
  created() {
    this.ALL = ALL;
    this.ACTIVE = ACTIVE;
    this.COMPLETED = COMPLETED;
  },
  mounted() {
    if (this.$route.path === "/") {
      this.linkSelected = ALL;
    } else {
      this.linkSelected = this.$route.path.substring(1);
    }
  },
};
</script>
<style>
.footer {
  color: #777;
  padding: 10px 15px;
  height: 20px;
  text-align: center;
  border-top: 1px solid #e6e6e6;
}

.footer:before {
  content: "";
  position: absolute;
  right: 0;
  bottom: 0;
  left: 0;
  height: 50px;
  overflow: hidden;
  box-shadow: 0 1px 1px rgba(0, 0, 0, 0.2), 0 8px 0 -3px #f6f6f6,
    0 9px 1px -3px rgba(0, 0, 0, 0.2), 0 16px 0 -6px #f6f6f6,
    0 17px 2px -6px rgba(0, 0, 0, 0.2);
}

.todo-count {
  float: left;
  text-align: left;
}

.todo-count strong {
  font-weight: 300;
}

.filters {
  margin: 0;
  padding: 0;
  list-style: none;
  position: absolute;
  right: 0;
  left: 0;
}

.filters li {
  display: inline;
}

.filters li a {
  color: inherit;
  margin: 3px;
  padding: 3px 7px;
  text-decoration: none;
  border: 1px solid transparent;
  border-radius: 3px;
}

.filters li a:hover {
  border-color: rgba(175, 47, 47, 0.1);
}

.filters li a.selected {
  border-color: rgba(175, 47, 47, 0.2);
}

.clear-completed,
.clear-completed:active {
  float: right;
  position: relative;
  line-height: 20px;
  text-decoration: none;
  cursor: pointer;
}

.clear-completed:hover {
  text-decoration: underline;
}

@media (max-width: 430px) {
  .footer {
    height: 50px;
  }

  .filters {
    bottom: 10px;
  }
}
</style>
