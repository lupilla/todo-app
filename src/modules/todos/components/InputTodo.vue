<template>
  <div id="newTodo">
    <input
      type="text"
      class="new-todo"
      v-model="name"
      autofocus
      placeholder="new todo..."
      aria-label="Add a new Todo"
      data-id-test="input"
      @keypress.enter="addNewTodo"
    />
  </div>
</template>
<script>
import { mapActions } from "vuex";

export default {
  name: "InputTodo",
  data() {
    return {
      name: "",
    };
  },
  methods: {
    ...mapActions("todos", ["addTodo"]),
    async addNewTodo() {
      if (this.name) {
        try {
          await this.addTodo(this.name);
          this.name = "";
        } catch (error) {
          console.error(error);
        }
      }
    },
  },
};
</script>

<style scoped lang="scss">
.new-todo {
  position: relative;
  margin: 0;
  width: 100%;
  font-size: 24px;
  font-family: inherit;
  font-weight: inherit;
  line-height: 1.4em;
  border: 0;
  color: inherit;
  padding: 6px;
  border: 1px solid #999;
  box-shadow: inset 0 -1px 5px 0 rgba(0, 0, 0, 0.2);
  box-sizing: border-box;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.new-todo {
  padding: 16px 16px 16px 60px;
  border: none;
  background: rgba(0, 0, 0, 0.003);
  box-shadow: inset 0 -2px 1px rgba(0, 0, 0, 0.03);
}
</style>
