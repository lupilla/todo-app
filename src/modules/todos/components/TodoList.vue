<template>
  <main id="main" class="main" v-if="showCheckAllTodos">
    <div>
      <input
        id="checkAll"
        type="checkbox"
        class="toggle-all"
        data-id-test="checkAll"
        aria-label="Check all Todos"
        @click="changeTodosStatus"
        v-model="selected"
      />
      <label for="checkAll"></label>
    </div>
    <ul class="todo-list">
      <li
        v-for="todo in todos"
        :key="todo.id"
        data-id-test="edit"
        :class="[{ editing: todo.isEditing }, { completed: todo.completed }]"
        @dblclick="changeEditMode(todo)"
      >
        <TodoItem
          :todo="todo"
          @edit="editTodo"
          @change-status="changeTodoStatus(todo)"
          @stop-edit="stopEditing(todo)"
        ></TodoItem>
      </li>
    </ul>
  </main>
</template>
<script>
import TodoItem from "@/modules/todos/components/TodoItem.vue";
import { mapGetters, mapActions } from "vuex";

export default {
  name: "TodoList",
  data() {
    return {
      selected: false,
    };
  },
  components: {
    TodoItem,
  },
  props: {
    todos: {
      type: Array,
      default: () => [],
    },
  },
  methods: {
    ...mapActions("todos", [
      "updateTodo",
      "changeCompleteTodo",
      "changeCompletedTodos",
      "changeEditableTodo",
      "changeEditableTodos",
    ]),
    async changeEditMode(todo) {
      await this.changeEditableTodo(todo);
    },
    async changeTodosStatus() {
      this.selected = !this.selected;
      try {
        await this.changeCompletedTodos(this.selected);
      } catch (error) {
        console.error(error);
      }
    },
    async changeTodoStatus(todo) {
      try {
        await this.changeCompleteTodo(todo);
        this.statusChanged();
      } catch (error) {
        console.error(error);
      }
    },
    async stopEditing(todo) {
      await this.changeEditableTodo(todo);
    },
    async editTodo(todo) {
      try {
        await this.updateTodo(todo);
        await this.changeEditableTodos();
        this.statusChanged();
      } catch (error) {
        console.error(error);
      }
    },
    statusChanged() {
      if (this.$route.name === "allTodos" && this.getCompletedLength) {
        this.selected = true;
      } else if (this.$route.name === "activeTodos") {
        this.selected = false;
      } else if (this.$route.name === "completedTodos") {
        this.selected = true;
      } else {
        this.selected = false;
      }
    },
  },
  computed: {
    ...mapGetters("todos", ["completedTodos"]),
    showCheckAllTodos() {
      return this.todos.length > 0;
    },
    getCompletedLength() {
      return this.completedTodos.length === this.todos.length;
    },
  },
  watch: {
    "$route.name": {
      handler: function () {
        if (this.todos) {
          this.statusChanged();
        }
      },
      deep: true,
      immediate: true,
    },
    todos: {
      handler: function () {
        if (this.todos) {
          this.statusChanged();
        }
      },
      deep: true,
      immediate: true,
    },
  },
};
</script>
<style>
.main {
  position: relative;
  z-index: 2;
  border-top: 1px solid #e6e6e6;
}

.todo-list {
  margin: 0;
  padding: 0;
  list-style: none;
}

.todo-list li {
  position: relative;
  font-size: 24px;
  border-bottom: 1px solid #ededed;
}

.todo-list li:last-child {
  border-bottom: none;
}

.todo-list li.editing {
  border-bottom: none;
  padding: 0;
}

.todo-list li.editing .edit {
  display: block;
  width: 506px;
  padding: 12px 16px;
  margin: 0 0 0 43px;
}

.todo-list li.editing .view {
  display: none;
}

.todo-list li label {
  word-break: break-all;
  padding: 15px 15px 15px 60px;
  display: block;
  line-height: 1.2;
  transition: color 0.4s;
}

.todo-list li.editing:last-child {
  margin-bottom: -1px;
}

.todo-list li:hover .destroy {
  display: block;
}

.toggle-all {
  text-align: center;
  border: none; /* Mobile Safari */
  opacity: 0;
  position: absolute;
}

.toggle-all + label {
  width: 60px;
  height: 34px;
  font-size: 0;
  position: absolute;
  top: -52px;
  left: -13px;
  -webkit-transform: rotate(90deg);
  transform: rotate(90deg);
}

.toggle-all + label:before {
  content: "‚ùØ";
  font-size: 22px;
  color: #e6e6e6;
  padding: 10px 27px 10px 27px;
}

.toggle-all:checked + label:before {
  color: #737373;
}

/*
	Hack to remove background from Mobile Safari.
	Can't use it globally since it destroys checkboxes in Firefox
*/
@media screen and (-webkit-min-device-pixel-ratio: 0) {
  .toggle-all,
  .todo-list li .toggle {
    background: none;
  }

  .todo-list li .toggle {
    height: 40px;
  }
}
</style>
