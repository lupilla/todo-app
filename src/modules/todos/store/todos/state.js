/**
 * General state object of todo module
 * It contains basically an array with the todos collection
 */
export default () => ({
  todos: [],
});
