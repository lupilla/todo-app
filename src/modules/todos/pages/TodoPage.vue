<template>
  <div id="todo-page">
    <h1 data-id-test="header">todos</h1>
    <InputTodo></InputTodo>
    <TodoList :todos="todoList"></TodoList>
    <FooterComponent :todos="todoList"></FooterComponent>
  </div>
</template>
<script>
import InputTodo from "@/modules/todos/components/InputTodo.vue";
import TodoList from "@/modules/todos/components/TodoList.vue";
import FooterComponent from "@/modules/todos/components/FooterComponent.vue";
import { mapActions, mapGetters } from "vuex";

export default {
  name: "TodoPage",
  components: {
    InputTodo,
    TodoList,
    FooterComponent,
  },
  computed: {
    ...mapGetters("todos", ["allTodos", "activeTodos", "completedTodos"]),
    todoList() {
      return this[this.$route.name];
    },
  },
  methods: {
    ...mapActions("todos", ["loadTodos", "addTodo"]),
    async getTodos() {
      try {
        await this.loadTodos();
      } catch (error) {
        console.error(error);
      }
    },
  },
  created() {
    this.getTodos();
  },
};
</script>

<style scoped lang="scss">
h1 {
  position: absolute;
  top: -155px;
  width: 100%;
  font-size: 100px;
  font-weight: 100;
  text-align: center;
  color: rgba(175, 47, 47, 0.15);
  -webkit-text-rendering: optimizeLegibility;
  -moz-text-rendering: optimizeLegibility;
  text-rendering: optimizeLegibility;
}
</style>
