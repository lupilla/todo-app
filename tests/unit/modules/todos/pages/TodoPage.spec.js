import { shallowMount } from "@vue/test-utils";
import TodoPage from "@/modules/todos/pages/TodoPage.vue";
import createVuexStore from "../../../mocks/mock-store.js";
import { INITIAL_STATE } from "../../../mocks/state";
import InputTodo from "@/modules/todos/components/InputTodo";
import TodoList from "@/modules/todos/components/TodoList";
import FooterComponent from "@/modules/todos/components/FooterComponent";
const store = createVuexStore();
describe("TodoPage tests", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(TodoPage, {
      global: {
        plugins: [store],
      },
      computed: {
        todoList: () => INITIAL_STATE.todos,
      },
    });
  });
  test("should match with snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("should show the proper components when todo list has entries", () => {
    const inputTodo = wrapper.findComponent(InputTodo);
    const todoList = wrapper.findComponent(TodoList);
    const footerComponent = wrapper.findComponent(FooterComponent);
    expect(inputTodo.exists()).toBe(true);
    expect(todoList.exists()).toBe(true);
    expect(footerComponent.exists()).toBe(true);
  });

  test("should show the proper components when todo list has not entries", () => {
    wrapper = shallowMount(TodoPage, {
      global: {
        plugins: [store],
      },
      computed: {
        todoList: () => [],
      },
    });
    const inputTodo = wrapper.findComponent(InputTodo);
    const todoList = wrapper.findComponent(TodoList);
    const footerComponent = wrapper.findComponent(FooterComponent);
    expect(inputTodo.exists()).toBe(true);
    expect(todoList.exists()).toBe(true);
    expect(footerComponent.exists()).toBe(true);
  });

  test("should show a header with 'todos' text inside", () => {
    const header = wrapper.find('[data-id-test="header"]');
    expect(header.exists()).toBe(true);
    expect(header.text()).toBe("todos");
  });
});

describe("TodoPage methods tests", () => {
  test("should call getTodos on mount hook", async () => {
    const wrapper = await shallowMount(TodoPage, {
      global: {
        plugins: [store],
      },
      computed: {
        todoList: () => INITIAL_STATE.todos,
      },
    });
    const spyGetTodos = jest.spyOn(wrapper.vm, "getTodos");
    const spyLoadTodos = jest.spyOn(wrapper.vm, "loadTodos");
    wrapper.vm.getTodos();
    expect(spyGetTodos).toHaveBeenCalled();
    expect(spyLoadTodos).toHaveBeenCalledTimes(1);
  });
});
