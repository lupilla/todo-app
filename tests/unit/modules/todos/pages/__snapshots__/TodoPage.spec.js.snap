// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoPage tests should match with snapshot 1`] = `
<div id="todo-page">
  <h1 data-id-test="header">todos</h1>
  <input-todo-stub></input-todo-stub>
  <todo-list-stub todos="[object Object],[object Object],[object Object]"></todo-list-stub>
  <footer-component-stub todos="[object Object],[object Object],[object Object]"></footer-component-stub>
</div>
`;
